### PR TITLE
Adds a rake command to import datasets ..

### LINKED ISSUE #29: Dataset import
Import legacy datasets into Beta.  Need to check models for the extended metadata where the dataset also has INSPIRE metadata attached.

### PR DESCRIPTION
**Pull request by rossjones**
.. from the legacy DGU data dump.

Fixes #29 

Commands should be run as (replacing the page to the file)

```
rake import:datasets[../data/datasets.jsonl]
```

This takes up to an hour to run, but doesn't need to be run often.

The jsonl file contains lots of entries, for example

```
{
  "author": null,
  "author_email": null,
  "creator_user_id": null,
  "extras": [
    {
      "id": "7a9f9a71-d855-4d50-b5a2-683557410f02",
      "key": "publish-date",
      "package_id": "67436432-07c3-4964-a365-5eb58d68a152",
      "revision_id": "5da82209-94f8-4e1f-be5e-add718e6c6c4",
      "revision_timestamp": "2013-08-31T00:56:15.435553",
      "state": "active",
      "value": "No Release Planned"
    },
    {
      "id": "7b9a0de1-bbe9-4505-b86b-3aded7c87d70",
      "key": "release-notes",
      "package_id": "67436432-07c3-4964-a365-5eb58d68a152",
      "revision_id": "5da82209-94f8-4e1f-be5e-add718e6c6c4",
      "revision_timestamp": "2013-08-31T00:56:15.435553",
      "state": "active",
      "value": "Personal Data"
    },
    {
      "id": "c6f57cd9-c708-4c9f-94bd-46159d9e3548",
      "key": "unpublished",
      "package_id": "67436432-07c3-4964-a365-5eb58d68a152",
      "revision_id": "5da82209-94f8-4e1f-be5e-add718e6c6c4",
      "revision_timestamp": "2013-08-31T00:56:15.435553",
      "state": "active",
      "value": "true"
    }
  ],
  "groups": [],
  "id": "67436432-07c3-4964-a365-5eb58d68a152",
  "isopen": false,
  "license_id": "",
  "license_title": "",
  "maintainer": null,
  "maintainer_email": null,
  "metadata_created": "2013-08-31T00:56:15.435553",
  "metadata_modified": "2013-08-31T00:56:15.435553",
  "name": "060-data-stats",
  "notes": "Ethnicity data",
  "num_resources": 0,
  "num_tags": 0,
  "organization": {
    "approval_status": "pending",
    "created": "2012-06-27T14:48:39.411478",
    "description": "We protect the security, independence and interests of our country at home and abroad. We work with our allies and partners whenever possible. Our aim is to ensure that the armed forces have the training, equipment and support necessary for their work, and that we keep within budget.\r\n\r\nMOD is a ministerial department, supported by 28 agencies and public bodies.\r\n\r\nhttps://www.gov.uk/government/organisations/ministry-of-defence\r\n\r\n",
    "id": "5db6e904-ea2f-42a7-93bd-a61da059246f",
    "image_url": "",
    "is_organization": true,
    "name": "ministry-of-defence",
    "revision_id": "b270ec2f-1112-4c1a-87fe-cd5a67f972e0",
    "revision_timestamp": "2013-11-05T07:07:17.647091",
    "state": "active",
    "title": "Ministry of Defence",
    "type": "organization"
  },
  "owner_org": "5db6e904-ea2f-42a7-93bd-a61da059246f",
  "private": false,
  "publish-date": "No Release Planned",
  "relationships_as_object": [],
  "relationships_as_subject": [],
  "release-notes": "Personal Data",
  "resources": [],
  "revision_id": "5da82209-94f8-4e1f-be5e-add718e6c6c4",
  "revision_timestamp": "2013-08-31T00:56:15.435553",
  "state": "active",
  "tags": [],
  "title": "060-Data-Stats",
  "tracking_summary": {
    "recent": 0,
    "total": 0
  },
  "type": "dataset",
  "unpublished": "true",
  "url": null,
  "version": null
}

```

### CODE REVIEW

File: lib/tasks/import.rake [hunk 1]
```diff
@@ -31,6 +32,97 @@ namespace :import do
...
+      d.published_date = obj["metadata_created"]
+      d.created_at = obj["metadata_created"]
+      d.updated_at = obj["metadata_modified"]
+      d.dataset_type = dataset_type(obj)
+      d.harvested = harvested?(obj)
+      d.location1 = ""
+      d.location2 = ""
+      d.location3 = ""
+      d.legacy_metadata = ""
+      d.save()
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: save! so it fails loudly

File: lib/tasks/import.rake [hunk 2]
```diff
@@ -31,6 +32,97 @@ namespace :import do
...
+        datafile.created_at = d.created_at
+        datafile.updated_at = d.updated_at
+
+        if resource["date"]
+          dates = get_start_end_date(resource["date"])
+          datafile.start_date = dates[0]
+          datafile.end_date   = dates[1]
+        end
+
+        datafile.save()
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: `save!`

File: lib/tasks/import.rake [hunk 3]
```diff
@@ -31,6 +32,97 @@ namespace :import do
...
+  inspire.import_source = get_extra(extras, 'import_source')
+  inspire.metadata_date = get_extra(extras, 'metadata-date')
+  inspire.metadata_language = get_extra(extras, 'metadata-language')
+  inspire.provider = get_extra(extras, 'provider')
+  inspire.resource_type = get_extra(extras, 'resource-type')
+  inspire.responsible_party = get_extra(extras, 'responsible-party')
+  inspire.spatial = get_extra(extras, 'spatial')
+  inspire.spatial_data_service_type = get_extra(extras, 'spatial-data-service-type')
+  inspire.spatial_reference_system = get_extra(extras, 'spatial-reference-system')
+  inspire
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: Inspire dataset doesn't get saved

File: lib/tasks/import.rake [hunk 4]
```diff
@@ -42,3 +134,93 @@ def json_from_lines(filename)
...
+  if parts and parts.length == 2
+    return calculate_dates_for_month(parts[0].to_i, parts[1].to_i)
+  end
+
+  ["", ""]
+end
+
+# Date helpers
+def calculate_dates_for_month(month, year)
+  days = Time.days_in_month(month, year)
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: There are helpers for `start_of_month` and `end_of_month`

File: lib/tasks/import.rake [hunk 5]
```diff
@@ -42,3 +134,93 @@ def json_from_lines(filename)
...
+#
+# {"id"=>"7096e248-1129-422f-a4b9-5ee570cd2f75",
+#  "key"=>"spatial-data-service-type",
+#  "package_id"=>"00b14406-02bf-4021-b538-8069223da623",
+#  "revision_id"=>"1e138922-8638-47b0-b15e-f87ff3b96f35",
+#  "revision_timestamp"=>"2016-06-16T08:43:55.184222",
+#  "state"=>"active",
+#  "value"=>""}
+def get_extra(extras, key)
+  return "" if !extras
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: nb you can also use `unless`

File: lib/tasks/import.rake [hunk 6]
```diff
@@ -31,6 +32,97 @@ namespace :import do
...
+
+        datafile.save()
+      end
+
+    end
+  end
+
+end
+
+def add_inspire_metadata(dataset_id, dataset)
```
Comment (govuklaurence) [inline lib/tasks/import.rake]: `dataset_id` is never used

File: lib/tasks/import.rake (hunk 3 above)
Comment (rossjones) [inline lib/tasks/import.rake]: It's saved by the caller.

File: lib/tasks/import.rake (hunk 4 above)
Comment (rossjones) [inline lib/tasks/import.rake]: We're only interested in how many days (the start'll always be 1, won't it?).

File: lib/tasks/import.rake (hunk 3 above)
Comment (govuklaurence) [inline lib/tasks/import.rake]: 👍

**Pull request merged**